pkg/auth/providers/oidc/oidc_provider.go

```diff
@@ -448,6 +458,71 @@ func (o *OpenIDCProvider) getUserInfo(ctx *context.Context, config *v32.OIDCConf
 	return userInfo, oauth2Token, nil
 }
 
+func (o *OpenIDCProvider) getClaimInfoFromToken(ctx *context.Context, config *v32.OIDCConfig, token *oauth2.Token, userName string) (*ClaimInfo, error) {
+	var userInfo *oidc.UserInfo
+	var err error
+	var claimInfo *ClaimInfo
+
+	updatedContext, err := AddCertKeyToContext(*ctx, config.Certificate, config.PrivateKey)
+	if err != nil {
+		return nil, err
+	}
+
+	provider, err := o.getOIDCProvider(updatedContext, config)
+	if err != nil {
+		return nil, err
+	}
+	oauthConfig := ConfigToOauthConfig(provider.Endpoint(), config)
+	var verifier = provider.Verifier(&oidc.Config{ClientID: config.ClientID})
+
+	// Valid will return false if access token is expired
+	if !token.Valid() {
+		// since token is not valid, the TokenSource func will attempt to refresh the access token
+		// if the refresh token has not expired
+		logrus.Debugf("[generic oidc] getUserInfo: attempting to refresh access token")
+		reusedToken, err := oauth2.ReuseTokenSource(token, oauthConfig.TokenSource(updatedContext, token)).Token()
+		if err != nil {
+			return nil, err
+		}
+		if !reflect.DeepEqual(token, reusedToken) {
+			err := o.UpdateToken(reusedToken, userName)
+			if err != nil {
+				return nil, fmt.Errorf("failed to update token: %w", err)
+			}
+		}
+		token = reusedToken
+	}
+
+	idToken, err := verifier.Verify(updatedContext, token.AccessToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify ID token: %w", err)
+	}
+	if err := idToken.Claims(&claimInfo); err != nil {
+		return nil, fmt.Errorf("failed to parse claims: %w", err)
+	}
+
+	if config.AcrValue != "" {
+		acrValue, err := parseACRFromAccessToken(token.AccessToken)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse ACR from access storedToken: %w", err)
+		}
+		if !isValidACR(acrValue, config.AcrValue) {
+			return nil, errors.New("failed to validate ACR")
+		}
+	}
+
+	logrus.Debugf("[generic oidc] getUserInfo: getting user info")
```

Comment:
I added the user ID to the log in the 2 places https://github.com/rancher/rancher/pull/48102/commits/7ecab38af09e08b8573eed6bbc54e9502b122082 

> It feels like we could fix this log entry to also include [generic oidc]

I left the prefix because we also have it in https://github.com/rancher/rancher/blob/main/pkg/auth/providers/oidc/oidc_provider.go#L439 , it looks like we add this prefix for other auth providers when getting the user info, but I would not add it to more places.

> (Maybe we could standardise the logging so that every log call doesn't include the prefix as part of the string?)

Absolutely, but probably out of scope for this PR